autonat/autonat-v2.md

```diff
@@ -0,0 +1,222 @@
+unattractive, the number of bytes is decided such that it's sufficiently larger
+than a new connection handshake cost.
+
+On receiving a `DialRequest`, the server selects the first address it is capable
+of dialing. If this selected address has a IP different from the client's
+observed ip, the server sends a `DialDataRequest` message with `numBytes` set to
+a sufficiently large value on the `/libp2p/autonat/2.0.0/dial-request` stream
+
+Upon receiving a `DialDataRequest` message, the client decides whether to accept
+or reject the cost of dial. If the client rejects the cost, the client resets
+the stream and the `DialRequest` is considered aborted. If the client accepts
+the cost, the client starts transferring `numBytes` bytes to the server. The
+server on receiving `numBytes` bytes proceeds to dial the candidate address. 
+
+## Implementation Suggestions
+
+For any given address, implementations SHOULD do the following
+- periodically recheck reachability status
+- query multiple peers to determine reachability
+
+The suggested heuristic for implementations is to consider an address reachable
+if more than 3 peers report a successful dial and to consider an address
+unreachable if more than 3 peers report unsuccessful dials. 
+
+Implementations are free to use different heuristics than this one
+
+Implementations SHOULD only verify reachability for private addresses as defined
+in [RFC 1918](https://datatracker.ietf.org/doc/html/rfc1918) with peers that are
+on the same subnet
+
+
+## RPC Messages
+
+All RPC messages sent over a stream are prefixed with the message length in
+bytes, encoded as an unsigned variable length integer as defined by the
+[multiformats unsigned-varint spec][uvarint-spec]. 
+
+All RPC messages on stream `/libp2p/autonat/2.0.0/dial` are of type
+`DialMessage`. A `DialRequest` message is sent as a `DialMessage` with the
+`dialRequest` field set to `DialRequest` message. `DialResponse` and
+`DialDataRequest` are handled similarly. On stream
+`/libp2p/autonat/2.0.0/attempt`, there is a single message type `AttemptMessage`
+where `DialAttempt` is handled similarly. 
+
+```proto
+syntax = "proto3";
+
+message DialMessage {
+  oneof msg {
+    DialRequest dialRequest   = 1;
+    DialResponse dialResponse = 2;
+    DialDataRequest dialDataRequest = 3;
+  }
+}
+
+message Candidate {
+  bytes addr = 1;
+  fixed64 nonce = 2;
+}
+
+message DialRequest {
+  repeated Candidate candidates = 1;
+}
+
+message DialDataRequest {
+  uint64 numBytes = 1;
+}
+
+message DialResponse {
+  enum ResponseStatus {
+      OK                        = 0;
+      E_DIAL_ERROR              = 100;
+      E_DIAL_REFUSED            = 101;
+      E_TRANSPORT_NOT_SUPPORTED = 102;
+      E_BAD_REQUEST             = 200;
+      E_INTERNAL_ERROR          = 300;
+  }
+
+  ResponseStatus status = 1;
+  int32 addrIdx = 2;
+}
+
+message AttemptMessage {
+  oneof msg {
+    DialAttempt dialAttempt = 1;
+  }
+}
+
+message DialAttempt {
+    fixed64 nonce = 1;
+}
```

Comment:
Does this structure provide us with much benefit? I am guessing the idea is evolvability. Now that we have a separate stream protocol for the attempt, we could also evolve things there if necessary (although at the cost of a roundtrip likely).

What do people think?

Reply:
Can you elaborate on what you mean by 
>Now that we have a separate stream protocol for the attempt, we could also evolve things there if necessary

Reply:
do you mean we can just send the nonce in the AttemptMessage?

Reply:
> Can you elaborate on what you mean by 
> >Now that we have a separate stream protocol for the attempt, we could also evolve things there if necessary

This message is only used on the `/libp2p/autonat/2.0.0/dial-attempt` protocol.

Thus, if we wanted to change the message, we could also just change to: `/libp2p/autonat/3.0.0/dial-attempt` instead of extending the `oneof` with a new message type.

Currently, the `oneof` introduces an additional indirection and error path in the code.

Reply:
> do you mean we can just send the nonce in the AttemptMessage? 

Yes.

Reply:
another great suggestion 🙇 

I have removed the AttemptMessage wrapper. 

I have also made the protocols `autonat/2/dial` and `autonat/2/attempt`. I don't see any benefit from using semver like protocol numbering here.